contracts/Migrator.sol

```diff
@@ -0,0 +1,239 @@
+        return _migrate(ogvAmount, msg.sender);
+    }
+
+    /**
+     * @notice Migrates OGV stakes to OGN. Can also include unstaked OGN & OGV
+     *          balances from the user's wallet (if specified).
+     * @param lockupIds OGV Lockup IDs to be migrated
+     * @param ogvAmountFromWallet Extra OGV balance from user's wallet to migrate & stake
+     * @param ognAmountFromWallet Extra OGN balance from user's wallet to stake
+     * @param migrateRewards If true, Migrate & Stake received rewards
+     * @param newStakeAmount Max amount of OGN (from wallet+unstake) to stake
+     * @param newStakeDuration Duration of the new stake
+     */
+    function migrate(
+        uint256[] calldata lockupIds,
+        uint256 ogvAmountFromWallet,
+        uint256 ognAmountFromWallet,
+        bool migrateRewards,
+        uint256 newStakeAmount,
+        uint256 newStakeDuration
+    ) external isSolvent {
+        if (!isMigrationActive()) {
+            revert MigrationIsInactive();
+        }
+
+        if (lockupIds.length == 0) {
+            revert LockupIdsRequired();
+        }
+
+        // Unstake
+        (uint256 ogvAmountUnlocked, uint256 rewardsCollected) = ogvStaking.unstakeFrom(msg.sender, lockupIds);
+
+        if (migrateRewards) {
+            // Include rewards if needed
+            ogvAmountFromWallet += rewardsCollected;
+        }
+
+        ogvAmountFromWallet += ogvAmountUnlocked;
+
+        if (ognAmountFromWallet > 0) {
+            // Transfer in additional OGN to stake from user's wallet
+            ogn.transferFrom(msg.sender, address(this), ognAmountFromWallet);
+        }
+
+        // Migrate OGV to OGN and include that along with existing balance
+        ognAmountFromWallet += _migrate(ogvAmountFromWallet, address(this));
+
+        if (ognAmountFromWallet < newStakeAmount) {
+            revert InvalidStakeAmount();
+        }
+
+        uint256 ognToWallet = ognAmountFromWallet - newStakeAmount;
+        if (ognToWallet > 0) {
+            ogn.transfer(msg.sender, ognToWallet);
+        }
+
+        if (newStakeAmount > 0) {
+            // Stake it
+            ognStaking.stake(
+                newStakeAmount,
+                newStakeDuration,
+                msg.sender,
+                false,
+                -1 // New stake
+            );
+        }
+
+        emit LockupsMigrated(msg.sender, lockupIds, newStakeAmount, newStakeDuration);
+    }
+
+    /**
+     * @notice Migrates caller's OGV to OGN and sends it to the `receiver`
+     * @return ognReceived OGN Received
+     */
+    function _migrate(uint256 ogvAmount, address receiver) internal returns (uint256 ognReceived) {
+        if (!isMigrationActive()) {
+            revert MigrationIsInactive();
+        }
+
+        ognReceived = (ogvAmount * CONVERSION_RATE) / 1 ether;
+
+        emit TokenExchanged(ogvAmount, ognReceived);
+
+        ogv.transferFrom(msg.sender, address(this), ogvAmount);
```

Comment:
With the user allowance, you could use `burnFrom` reducing slightly the gas usage.

Reply:
The actual spec was to not burn any tokens until the contract is decommissioned

Reply:
Given that we've changed to running the checks off the ogv totalSupply instead of balances, we could switch to just burning ogv as it comes in. I would be glad to do this.

Reply:
This would also get rid of the need to account for ogv held by the contract in the solvency check.